pkg/operators/limiter/limiter.go

```diff
@@ -0,0 +1,164 @@
+// this operator is when you are already sorting data within an array of data
+// and you want to filter out the top `X` entries.
+package limiter
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	name            = "limiter"
+	ParamMaxEntries = "max-entries"
+	Priority        = 9600
+)
+
+type limiterOperator struct{}
+
+func (l *limiterOperator) Name() string {
+	return name
+}
+
+func (l *limiterOperator) Init(params *params.Params) error {
+	return nil
+}
+
+func (l *limiterOperator) GlobalParams() api.Params {
+	return nil
+}
+
+func (l *limiterOperator) InstanceParams() api.Params {
+	return api.Params{
+		{
+			Key:   ParamMaxEntries,
+			Title: "Max Entries",
+			Description: "The maximal number of entries for each batch of data. " +
+				"If using multiple data sources, prefix the value with 'datasourcename:' and separate with ';'.",
+			DefaultValue: "50",
+			TypeHint:     api.TypeString,
+		},
+	}
+}
+
+func (l *limiterOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, instanceParamValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	for _, ds := range gadgetCtx.GetDataSources() {
+		if ds.Type() == datasource.TypeArray {
+			return &limiterOperatorInstance{
+				maxEntries: instanceParamValues[ParamMaxEntries],
+			}, nil
+		}
+	}
+	return nil, nil
+}
+
+func (l *limiterOperator) Priority() int {
+	return Priority
+}
+
+type limiterOperatorInstance struct {
+	maxEntries string
+}
+
+func (l *limiterOperatorInstance) Name() string {
+	return name
+}
+
+// parseMaxEntries parses the max-entries parameter and returns a map of data
+// source names to the maximum number of entries to keep. If no data source is
+// specified, the value is used for all data sources.
+func parseMaxEntries(maxEntries string) (map[string]int, error) {
+	dsLimiters := make(map[string]int)
+	for _, s := range strings.Split(maxEntries, ";") {
+		dsMaxEntries := strings.Split(s, ":")
+		dsName := ""
+		maxEntriesStr := dsMaxEntries[0]
+		if len(dsMaxEntries) == 2 {
+			dsName = dsMaxEntries[0]
+			maxEntriesStr = dsMaxEntries[1]
+		}
+		val, err := strconv.Atoi(maxEntriesStr)
+		if err != nil {
+			return nil, fmt.Errorf("atoi %q: %w", maxEntriesStr, err)
+		}
+		if val < 0 {
+			return nil, fmt.Errorf("invalid value %d for %s", val, ParamMaxEntries)
+		}
+		dsLimiters[dsName] = val
+	}
+
+	// Check edge cases
+	if _, ok := dsLimiters[""]; ok {
+		if len(dsLimiters) > 1 {
+			return nil, fmt.Errorf("mixing %s rules with and without specifying data source", ParamMaxEntries)
+		}
+	}
+
+	return dsLimiters, nil
+}
+
+func (l *limiterOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error {
+	dsLimiters, err := parseMaxEntries(l.maxEntries)
```

Comment:
Maybe move the validation / parsing to InstantiateDataOperator, so that errors can be caught early.

Reply:
Let's continue the discussion in https://github.com/inspektor-gadget/inspektor-gadget/pull/3327#discussion_r1726835433